ros2cli/ros2cli/daemon/daemonize.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2021 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copyreg
+import os
+import pickle
+import platform
+import socket
+import subprocess
+import sys
+import threading
+
+from ros2cli.helpers import wait_for
+
+
+class PicklerForProcess(pickle.Pickler):
+
+    def __init__(self, process, *args, **kwargs):
+        super().__init__(process.stdin, *args, **kwargs)
+        self.process = process
+        self.dispatch_table = copyreg.dispatch_table.copy()
+        self.dispatch_table[socket.socket] = self.reduce_socket
+        self.dispatch_table[threading.Event] = self.reduce_event
+
+    def reduce_event(self, obj):
```

Comment:
why is this method needed?

Reply:
Good question. When pickling a `LocalXMLRPCServer` instance, eventually we hit a `threading.Event` attribute in [`socketserver.BaseServer`](https://github.com/python/cpython/blob/5fc784e28ab9cd02c30395d1973ff825b787efa2/Lib/socketserver.py#L204). Thus this method, which simply makes sure that another `threading.Event` instance is constructed on the other end.

Reply:
If in the future `socketserver.BaseServer` implementation change, do we also need to change this class?
That sounds a bit fragile.
Could we send something simpler?
Like only the socket object, and then creating a new XMLRPC server based on that socket (if that's possible).

Reply:
> If in the future `socketserver.BaseServer` implementation change, do we also need to change this class?
> That sounds a bit fragile.

That is correct. That piece of code's been there unchanged for 14 years though. I wouldn't expect it to change anytime soon, and if it does, it doesn't necessarily mean we have to change this code -- only if it changes in a way that requires the `Pickler` subclass to be extended somehow.

> Could we send something simpler?
> Like only the socket object, and then creating a new XMLRPC server based on that socket (if that's possible).

I thought about it. It is possible by re-implementing [`socketserver.TCPServer`](https://github.com/python/cpython/blob/5fc784e28ab9cd02c30395d1973ff825b787efa2/Lib/socketserver.py#L390) and [`xmlrpc.server.SimpleXMLRPCServer`](https://github.com/python/cpython/blob/5fc784e28ab9cd02c30395d1973ff825b787efa2/Lib/xmlrpc/server.py#L580) classes here. I don't love duplicating but that'd do.

Reply:
>  only if it changes in a way that requires the Pickler subclass to be extended somehow.

Yes, that's true.
I guess that would be unusual.

> I thought about it. It is possible by re-implementing socketserver.TCPServer and xmlrpc.server.SimpleXMLRPCServer classes here. I don't love duplicating but that'd do.

Okay, if the existing objects don't support that I guess it's fair to serialize everything.

---

I wonder if there's any race when serializing `LocalXMLRPCServer`.
e.g. might the `LocalXMLRPCServer` instance be processing a request (and thus, modifying its internal state) while being serialized?
I haven't checked the python library code, but that would be a problem.

Reply:
> I wonder if there's any race when serializing `LocalXMLRPCServer` e.g. might the `LocalXMLRPCServer` instance be processing a request (and thus, modifying its internal state) while being serialized?

Good question. It cannot process a request until [`handle_request()`](https://github.com/python/cpython/blob/5fc784e28ab9cd02c30395d1973ff825b787efa2/Lib/socketserver.py#L273) is called so that's not a problem BUT if a connection request arrives during the brief period of time in which the socket exists in both processes, I don't know where that connection request will end up. 

That's likely OS specific. I haven't read what's up with file descriptor inheritance on Unix and handle duplication on Windows in that specific case. But in the worst case scenario I guess some client will get their connection refused when the parent process destroys the server.